src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Debugging/RemoteDebugInfoService.cs

```diff
@@ -113,27 +113,48 @@ protected override IRemoteDebugInfoService CreateService(in ServiceArgs args)
 
     private bool TryGetUsableProjectedIndex(RazorCodeDocument codeDocument, LinePosition hostDocumentPosition, out int projectedIndex)
     {
-        var hostDocumentIndex = codeDocument.Source.Text.GetPosition(hostDocumentPosition);
-
         projectedIndex = 0;
-        var languageKind = codeDocument.GetLanguageKind(hostDocumentIndex, rightAssociative: false);
-        // For C#, we just map
-        if (languageKind == RazorLanguageKind.CSharp &&
-            !_documentMappingService.TryMapToCSharpDocumentPosition(codeDocument.GetRequiredCSharpDocument(), hostDocumentIndex, out _, out projectedIndex))
-        {
-            return false;
-        }
-        // Otherwise see if there is more C# on the line to map to. This is for situations like "$$<p>@DateTime.Now</p>"
-        else if (languageKind == RazorLanguageKind.Html &&
-            !_documentMappingService.TryMapToCSharpPositionOrNext(codeDocument.GetRequiredCSharpDocument(), hostDocumentIndex, out _, out projectedIndex))
-        {
-            return false;
-        }
-        else if (languageKind == RazorLanguageKind.Razor)
+
+        var sourceText = codeDocument.Source.Text;
+        var hostDocumentIndex = sourceText.GetPosition(hostDocumentPosition);
+        var syntaxRoot = codeDocument.GetRequiredSyntaxRoot();
+        var csharpDocument = codeDocument.GetRequiredCSharpDocument();
+
+        // We want to find a position that maps to C# on the same line as the original request, but we might have to skip over
+        // some Razor/HTML nodes to find valid C#.
+        while (sourceText.GetLinePosition(hostDocumentIndex).Line == hostDocumentPosition.Line)
         {
-            return false;
+            if (_documentMappingService.TryMapToCSharpPositionOrNext(csharpDocument, hostDocumentIndex, out _, out projectedIndex))
```

Comment:
>TryMapToCSharpPositionOrNext [](http://example.com/codeflow?start=40&length=28)

not related to this PR: 

Hadn't seen this method before so I took a peek, and I'm curious whether TryMapToCSharpDocumentPositionInternal is missing a break statement when 

originalAbsoluteIndex > razorIndex && !nextCSharpPositionOnFailure

Reply:
Ahh, turns out we've discussed this before: https://github.com/dotnet/razor/pull/9291

TL;DR, source mappings are ordered by the GeneratedSpan (as the compiler adds source mappings as it writes the generated file, not necessarily as it reads the Razor file) so an early exit there would potentially miss something.

Reply:
OK, that makes sense. But now I'm confused how that code knows it's the "next" one if there are multiple C# ranges on the same line? The "next" code short circuits based on a mapping being on the line, not that it is the first one on that line after the span.